### DISCUSSION TITLE
The best way to have app-wide CMS'd data on a static site

### DISCUSSION THREAD

**Discussion by andrewfairlie** [vercel/next.js#17888] [Help]
What is the best way to load in app-wide CMS'd or API'd data (such as the navigation, some global SEO meta, etc.), ideally in `_app.js` so it can power things that persist between routes (such as a persistent navigation).

# Approach 1
- Fetch the data in `getStaticProps` and then pass it into a `<Layout>` component or similar
- Downside: you lose the persistence, because the navigation is within a layout component put on the page itself, and so changes each route change.

# Approach 2
- Fetch the data in `getInitialProps` on `_app.js` and pass it down as props to things like `<SiteHeader>` directly in `_app.js`
- Advantage: things like site headers can persist across page loads
- Disadvantage: I don't know the full ramifications of this, when I set `getInitialProps` I lose the bottom-right dev indicator, which makes me think things I'm unaware of are happening that might backfire on me

# Other approaches?

Or are there any other approaches to solve this problem that I've not considered here?

**Comment by thebuilder** [comment]
* Fetch shared data with `getStaticProps` on all pages.
* Read it from `pageProps` in `_app.js` (https://nextjs.org/docs/advanced-features/custom-app)

The disadvantage is you include the global layout/seo in the JSON for all pages. You'll need to rebuild the pages if anything global changes.

**Reply by andrewfairlie** [reply]
This looks promising, I'll try this out tomorrow and report back. Thanks for your input 👍

**Reply by moop-moop** [reply]
> * Fetch shared data with `getStaticProps` on all pages.
> * Read it from `pageProps` in `_app.js` (https://nextjs.org/docs/advanced-features/custom-app)
> 
> The disadvantage is you include the global layout/seo in the JSON for all pages. You'll need to rebuild the pages if anything global changes.

Can you elaborate with how this would be implemented?

**Comment by jurajkrivda** [comment]
Use getServerSideProps instead of getInitialProps.
Maybe you could use async function in next.config.

**Reply by andrewfairlie** [reply]
Thank you. I've carved some time out to play with this and I'll get it a go. Thanks for your help 👍

**Comment by hsdonkin** [comment]
We've had some luck with `next-plugin-preval` which runs at build time and lets you import data like a JSON

https://www.npmjs.com/package/next-plugin-preval
